extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js

```diff
@@ -695,16 +790,218 @@ export class AmpLightboxGallery extends AMP.BaseElement {
+      transform: `scale(${scale}) translate(${deltaX}px, ${deltaY}px)`,
+      transition: `${duration}ms transform ${SWIPE_TO_CLOSE_MOMENTUM_TIMING}`,
+    });
+
+    return delayAfterDeferringToEventLoop(this.win, duration);
+  }
+
+  /**
+   * Snaps back to the starting point, with the duration based on the distance
+   * that needs to be travelled.
+   * @param {number} finalDistance
+   * @return {Promise} A Promise that resolves once the snapping has completed.
+   * @private
+   */
+  snapBackFromSwipe_(finalDistance) {
+    const duration = finalDistance * SWIPE_TO_CLOSE_SNAP_BACK_TIME_FACTOR;
+
+    return this.mutateElement(() => {
+      setStyles(devAssert(this.carousel_), {
+        transform: '',
+        transition: `${duration}ms transform ease-out`,
+      });
+      setStyles(devAssert(this.mask_), {
+        opacity: '',
+        transition: `${duration}ms opacity ease-out`,
+      });
+      setStyles(devAssert(this.controlsContainer_), {
+        opacity: '',
+        transition: `${duration}ms opacity ease-out`,
+      });
+    }).then(() => {
+      return delayAfterDeferringToEventLoop(this.win, duration);
+    }).then(() => {
+      return this.mutateElement(() => {});
+    });
+  }
+
+  /**
+   * Adjusts the UI elements for the current swipe position in a swipe to
+   * dismiss gesture. This should be called in a mutate context.
+   * @param {!Element} carousel The carousel element to adjust. This is passed
+   *    as `this.carousel_` will be null if this is called after the lightbox
+   *    has been closed.
+   * @param {string} carouselTransform How to transform the carousel.
+   * @param {number|string} maskOpacity The opacity for the mask element.
+   * @param {number|string} controlsOpacity The opacity for the controls
+   *    container.
+   * @private
+   */
+  adjustForSwipePosition_(
+    carousel, carouselTransform, maskOpacity, controlsOpacity) {
+    setStyles(carousel, {
+      transform: carouselTransform,
+      transition: '',
+    });
+    setStyles(devAssert(this.mask_), {
+      opacity: maskOpacity,
+      transition: '',
+    });
+    setStyles(devAssert(this.controlsContainer_), {
+      opacity: controlsOpacity,
+      transition: '',
+    });
+  }
+
+  /**
+   * Releases the user's swipe to dismiss gesture. This carries the momentum
+   * forwards and either closes the lightbox or snaps back based on the speed
+   * and distance. This should be called in a mutate context.
+   * @param {number} scale The scale when releasing the swipe. We do not change
+   *    the scale as we carry forward any momentum.
+   * @param {number} velocityX The X velocity when the swipe was released.
+   * @param {number} velocityY The Y velocity when the swipe was released.
+   * @param {number} deltaX The x distance when the swipe was released.
+   * @param {number} deltaY The y distance when the swipe was released.
+   * @return {Promise} A Promise that resolves once the release is completed,
```

Comment:
nit: I think the type here is `!Promise`?

Reply:
Done.